list2/ex32/index.js

```diff
@@ -0,0 +1,38 @@
+import { generateMatrix, printMatrix } from "../helper.js";
+
+export function matrixRowNormalization() {
+  const rowCount = 12;
+  const columnCount = 13;
+
+  const matrix = generateMatrix({
+    rowCount,
+    columnCount,
+  });
+
+  printMatrix(matrix);
+  console.log("---------------");
+
+  function newMatrix({ matrix, rowCount, columnCount }) {
+    const modifiedMatrix = JSON.parse(JSON.stringify(matrix));
+
+    for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
+      const biggestValue = Math.max(...modifiedMatrix[rowIndex]);
+      console.log(`The biggest ${rowIndex + 1} row number: ${biggestValue}`);
+
+      for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
+        modifiedMatrix[rowIndex][columnIndex] = Math.round((matrix[rowIndex][columnIndex] / biggestValue) * 100) / 100;
+      }
+    }
+    return modifiedMatrix;
```

Comment:
You can avoid `JSON.parse(JSON.stringify(matrix))` by simply using `Array.prototype.map` to map the original matrix to the new one.

Reply:
Is that the same as `cosnt modifiedMatrix = [...matrix]` ?
If yes, what is the better choice?

Reply:
The three can be used to achieve the same result but they are fundamentally different.

1. `JSON.parse(JSON.stringify(matrix))`: This is a hacky way to create a copy of an object in Javascript. But it can lead to several problems, especially with complex objects that contain functions or Date objects for example, therefore, it is considered bad practice. 

2. `Array.prototype.map`: Is used to create new arrays with the same length of the original one but with a different inner structure. The image below visually explains some of the main `Array.prototype` functions, check to see if it helps your understanding:
![image](https://github.com/user-attachments/assets/ce40dc24-dc89-47e6-8178-4710b1da4154)

3. `[...matrix]`: This is known as the spread operator and can also be used to create copies of arrays, but we should always try to avoid copying arrays to mutate them afterward. Functional programming suggests that we should never change data after it is created. Instead of modifying existing data, new data structures are returned. For example:
```js
const person = {
  name: 'John Doe',
  age: 25,
  email: 'john@doe.com',
};

/** Mutating the object */
person.age = 26;
/** Using spread operator to create a new object and follow functional programming immutability principle */
const updatedPerson = { ...person, age: 26 };

/** The same goes for arrays */
const numbers = [1, 2, 3];
/** Mutating the array */
numbers.push(4);
/** Using spread operator to create a new array and follow functional programming immutability principle */
const updatedNumbers = [...numbers, 4];
```
I hope this heps!

Reply:
I'll correct the exercise again, now I think I understand

Reply:
`const matrixFiltered = matrix.map((row) => [...row]);`

Like that?